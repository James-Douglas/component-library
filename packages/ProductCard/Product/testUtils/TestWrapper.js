import React from 'react';
import ProductProvider from '../../Context/ProductProvider';

const ProductTestWrapper = ({
  // eslint-disable-next-line react/prop-types
  children, product, onDetails = () => {}, onSelect = () => {},
}) => (
  <ProductProvider product={product} onDetails={onDetails} onSelect={onSelect}>
    { children }
  </ProductProvider>
);

export default ProductTestWrapper;
