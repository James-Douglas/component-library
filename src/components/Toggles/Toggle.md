# Toggle component
The CTM Toggle component. Can render a variety of styles of toggle:

* Square (text only, <= 25 chars)
* Rectangle (text only, > 25 chars)
* Icon (square toggle with an icon)
* Picture (square toggle with a picture) 

## Usage
~~~js
<Toggle
  id={id}
  label={label}
  type={type}
  value={value}
  selectedValue={selectedValue}
  name={name}
  invalid={invalid}
  disabled={disabled}
  onToggle={onToggle}
  icon={icon}
  iconSize={iconSize}
  pictureOptions={pictureOptions}
  rectOptions={rectOptions}
/>
~~~

## Props
The ToggleGroup component accepts the following props:

- `id` *(required) string*
    * Unique identifier for the toggle
    
- `label` *(required) string*
    * Label for the toggle.
    
- `type` *(optional) string*
    * The type of toggle to render (either 'square' or 'rectangle').
    
-  `value` *(optional) string*
    * The value to be applied to the input field.
    
-  `selectedValue` *(optional) string*
    * The value of the currently selected (toggled on) toggle.
    
-  `isPrefill` *(optional) bool*
    * Whether the selected value is a prefill (e.g. auto-populated from a prior journey)
     
-  `name` *(optional) string*
    * The name to be applied to the input field.
    
-  `invalid` *(optional) boolean*
    * Applies invalid styling when true.
    
-  `disabled` *(optional) boolean*
    * Disables the toggle when true.
    
- `onToggle` *(optional) function*
    * Handler function called when a toggle is toggled on (called with object: { id, value } of the toggle).
    
- `icon` *(optional) string*
    * Name of the icon to be rendered on the toggle. (Note if a pictureOptions object is also passed this prop will be overridden).
    
- `iconSize` *(optional) number*
    * Size of the icon to be rendered on the toggle.
    
- `pictureOptions` *(optional) object*
    * Options object for rendering a picture on a toggle, attributes match those of the Picture component.
    
- `rectOptions` *(optional) object*
    * Options object for rendering rectangular toggles, attributes:
        * align ('center' (default), 'left', 'right') - text alignment
        * col (1,2,3,4,5,6,7,8,10,11,12) - width of toggle
        * height (number) - height (rem) of the toggle
