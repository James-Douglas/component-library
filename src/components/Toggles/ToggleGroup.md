# ToggleGroup component
The CTM ToggleGroup component, wraps a given set of toggles.

## Usage
~~~js
<ToggleGroup 
  label={label}
  id={id}
  name={name}
  tooltip={tooltip}
  handleChange={handleChange}
  selectedValue={selectedValue}
>
  <Toggle />
  <Toggle />
  <Toggle />
</ToggleGroup>
~~~

*Priority: picture > icon

## Props
The ToggleGroup component accepts the following props:

1. `name` *(required) string*
    * Name property to be passed to the toggles - required for radio buttons

7. `onToggle` *(required) function*
    * onChange handler function, called on select of a toggle with { id: <selected toggle id>, value: <selected toggle value> }

2. `label` *(optional) string*
    * Label for the ToggleGroup. Defaults to an empty string.

3. `tooltip` *(optional) object*
   * Tooltip object (see Tooltip documentation)
       
4. `forceFullWidth` *(optional) boolean* 
   * Forces the ToggleGroup to expand to 12 columns (default true for ToggleGroup)

5. `validationMessage` *(optional string)*
    * Displays given validation message and invalid styles on the component when provided.  
    
6. `id` *(optional) string*
    * Unique identifier for the toggle group
    
8. `rectOptions` *(optional) object*
    * Options object for rectangular toggles (see Toggle documentation)
    
9. `selectedValue` *(optional) string/number*
    * Value of the currently selected toggle (use to pre-select)
    
