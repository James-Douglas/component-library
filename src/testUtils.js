// test-utils.js
import {
  render, fireEvent, wait, act,
} from '@testing-library/react';
import ManorProvider from './providers/ManorProvider';

const customRender = (ui, options) => render(ui, { wrapper: ManorProvider, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export {
  customRender as render,
  // export other used functions to appease eslint
  fireEvent,
  wait,
  act,
};
