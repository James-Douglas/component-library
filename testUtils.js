import React from 'react';
import {
  render, fireEvent, wait, act,
} from '@testing-library/react';
import ManorProvider from './packages/Provider/ManorProvider';

const ManorTestWrapper = ({ children }) => (
  <ManorProvider>
    {children}
  </ManorProvider>
);

const customRender = (ui, options) => render(ui, { wrapper: ManorTestWrapper, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export {
  customRender as render,
  // export other used functions to appease eslint
  fireEvent,
  wait,
  act,
};
