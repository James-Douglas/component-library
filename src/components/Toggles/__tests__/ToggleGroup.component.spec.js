import React from 'react';
import { render, fireEvent } from '../../../testUtils';
import ToggleGroup, { getType, getChildren } from '../ToggleGroup.component';
import Toggle from '../Toggle.component';
import 'jest-styled-components';

describe('getType()', () => {
  it('returns square when all labels have length <= 25 chars', () => {
    const children = [
      { label: 'test a' },
      { label: 'test b' },
      { label: 'test c' },
    ];
    expect(getType(children)).toBe('square');
  });

  it('returns rectangle when a label has length > 25 chars', () => {
    const children = [
      { props: { label: 'test a' } },
      { props: { label: 'test b' } },
      { props: { label: 'test ccccccccccccccccccccc' } },
    ];
    expect(getType(children)).toBe('rectangle');
  });
});

describe('getChildren()', () => {
  let tempID = 'temp-id';
  const getTestChildren = (data) => data.map((child) => React.cloneElement(<Toggle id={tempID} label="test-label" value="test-value" />, { ...child }));
  const didToggleCb = jest.fn();

  it('returns cloned children with base props added', () => {
    const testChildren = getTestChildren([
      {
        label: 'test toggle a',
        id: 'a',
        value: 'val-a',
      },
    ]);
    const result = getChildren(testChildren, 'square', 'test-group', null, didToggleCb, null);
    const {
      name, id, selectedValue, handleToggle, type, value,
    } = result[0].props;

    expect(name).toEqual('test-group');
    expect(value).toEqual('val-a');
    expect(id).toEqual('a');
    expect(selectedValue).toEqual(null);
    expect(handleToggle).toEqual(didToggleCb);
    expect(type).toEqual('square');
  });

  it('adds id if none provided', () => {
    tempID = '';
    const testChildren = getTestChildren([
      {
        label: 'test toggle a',
      },
    ]);
    const result = getChildren(testChildren, 'square', 'test-group', null, didToggleCb, null);
    const { id } = result[0].props;
    expect(id).toEqual('toggle-0');
  });

  it('adds rectOptions if required', () => {
    const expectedRectOptions = { align: 'left' };
    const testChildren = getTestChildren([
      {
        label: 'test toggle a',
      },
    ]);
    const result = getChildren(testChildren, 'rectangle', 'test-group', null, didToggleCb, expectedRectOptions);
    const { rectOptions } = result[0].props;
    expect(rectOptions).toEqual(expectedRectOptions);
  });
});


describe('ToggleGroup', () => {
  it('renders with minimal props', () => {
    const { container } = render(<ToggleGroup
      name="test-toggle-group"
      handleToggle={() => {}}
    />);
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('calls handleToggle on toggle selection', () => {
    const onToggleCb = jest.fn();
    const { container } = render(
      <ToggleGroup name="test-toggle-group-b" handleToggle={onToggleCb}>
        <Toggle label="test toggle a" value="a" id="a" />
        <Toggle label="test toggle b" value="b" id="b" />
      </ToggleGroup>,
    );
    const toggleA = container.querySelector('#a');
    fireEvent.click(toggleA);
    expect(onToggleCb).toHaveBeenCalled();
    expect(onToggleCb.mock.calls[0][0]).toEqual('a');
  });

  it('adds justifyEnd property to tooltip when exists', () => {
    const handleChangeCb = jest.fn();
    const tooltip = { title: 'test' };
    const { container } = render(
      <ToggleGroup name="test-toggle-group-b" handleToggle={handleChangeCb} tooltip={tooltip}>
        <Toggle label="test toggle a" id="a" value="val-1" />
        <Toggle label="test toggle b" id="b" value="val-2" />
      </ToggleGroup>,
    );
    const tooltipWrapper = container.querySelector('[role="tooltip"]');
    expect(tooltipWrapper.parentNode).toHaveStyleRule('justify-content', 'flex-end');
  });
});
