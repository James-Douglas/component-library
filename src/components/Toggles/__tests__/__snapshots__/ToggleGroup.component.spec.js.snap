// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleGroup renders with minimal props 1`] = `"<label class=\\"sc-AxhCb hVSPzA label\\"></label><div class=\\"sc-AxheI hvWqxE\\"></div><div class=\\"sc-Axmtr fUrbvH\\"></div>"`;
